Allow confirming email in a different browser

monfresh opened the pull request:
**Why**: Some people might visit the site via a Service Provider (SP)
in one browser, but open the email confirmation link in a different
browser. This means that anything that was previously stored in the
session in the first browser won't be available in the second browser.
The consequence is that the user won't be redirected back to the SP
after they finish creating their account.

**How**:
- Add a new ServiceProviderRequest DB table

When a request is made from an SP, whether it's via SAML or OpenID
Connect, we create a new ServiceProviderRequest with the following
attributes: the full request URL, a random unique UUID (because it's
not guaranteed that separate requests will have a unique identifier
attached to them), the issuer, and LOA. We also store these same
attributes in the session to support the sign in scenario (as opposed
to account creation).

Then, we check to see if the user is fully authenticated, and if not,
we redirect them to the `/sign_up/start` page, and include the
request_id in the params, as well as in the links on that page that
point to the account creation and sign in pages. Note that I moved the
logic that determines whether or not the landing page should appear,
from the sessions controller to the SAML and OpenID Connect controllers,
since they are the ones that know where the user should go. This adheres
to the Tell, Don't Ask principle, and allows us to remove the
`show_start_page` key from the session.

When the user chooses to create an account, the form contains the
request_id in a hidden field so it can pass it on to the registrations
controller. To include the request_id in the confirmation email, I had
to create a new method based on the one Devise uses, but with the
ability to pass in the request_id. Since Devise automatically sends the
confirmation email when the user is created (via a callback), I had to
override the original method to do nothing.

By having the request_id in the email, that means that if the user
visits the confirmation URL in a different browser, we can go back to
storing the SP info in the session by looking up the
ServiceProviderRequest whose uuid matches the `_request_id` parameter
from the URL in the email. By going back to using the session from then
on, it frees us from having to keep passing in the request_id in URLs
and in forms. This is done in
`EmailConfirmationsController#process_valid_confirmation_token`.

Note that the request_id is prefixed with an underscore in the email to
make sure that the URL ends with the confirmation token. By default,
Rails sorts params in `link_to` helpers alphabetically, but we don't
want `request_id` to come last because if the user chooses to copy the
link instead of clicking it, and if they don't copy it correctly, the
app won't find the ServiceProviderRequest, whereas an invalid
confirmation token will display an error to the user.

Once the user finishes creating their account, when they click the
button to continue to the SP, the app makes the original SP request
again, calling the `before_action`s in the SAML and OpenId Connect
controllers a second time. However, we don't want to create a new
ServiceProviderRequest because we want to be able to delete the
original one after the user goes back to the SP. We don't want the
ServiceProviderRequests table to grow indefinitely. The only way
to know for sure whether or not a ServiceProviderRequest that matches
the request URL already exists is to query on its `url` field. The
problem is, given that we want to index that field since it will be
queried on every SP request, the url field value is too big for the
Postgres btree index. There are solutions to this problem (that I
haven't implemented before), but I didn't feel it was worth the
trouble when we can rely on the presence of `sp_session[:request_id]`
to determine whether or not a new ServiceProviderRequest should be
created or not. We also use the `sp_session[:request_id]` to determine
which ServiceProviderRequest to delete by find a matching uuid. Then,
we can safely delete the `:sp` Hash from the session.

By keeping track of the SP via the request_id, we can also remove the
session timeout code and flash message that dealt with the scenario
where a user makes a request from an SP, but then remains on the
sign in page for longer than 8 minutes. Note that all this does is
preserve the branded experience on the sign in page. Allowing the
SP info to be restored after sign in will be implemented in a
separate PR.

Note that this doesn't include the same fix for the scenario where a
user opens the password reset link in a different browser. That will be
in a separate PR.

zachmargolis commented on app/controllers/openid_connect/authorization_controller.rb:
```diff
@@ -47,6 +45,23 @@ def destroy
+                            errors: @authorize_form.errors.messages)
+    end
+
+    def track_create_action_analytics(result)
```
is this method extraction to reduce the ABC size of the `#create` method?

monfresh replied:
Correct.

zachmargolis commented on app/controllers/saml_idp_controller.rb:
```diff
@@ -38,6 +36,22 @@ def logout
+
+    yield true if needs_idv
+
+    yield false
```
this will `yield` twice when true, is that intentional?

monfresh replied:
That was not intentional, but it doesn't seem to cause a problem. However, I'll fix it so that it reads `yield needs_idv`. That's what I get for coding late at night.

zachmargolis commented on app/models/user.rb:
```diff
@@ -133,4 +133,17 @@ def downcase_keys
+  end
+
+  def send_custom_confirmation_instructions(id = nil)
+    generate_confirmation_token! unless @raw_confirmation_token
```
Where does `@ raw_confirmation_token` get set? Where does `generate_confirmation_token!` come from? Devise?

monfresh replied:
Yes, I copied Devise's `send_confirmation_instructions` with the only modification being the ability to pass in the request_id to the mailer. I'll add a comment.

zachmargolis commented on app/models/service_provider_request.rb:
```diff
@@ -0,0 +1,5 @@
+class ServiceProviderRequest < ActiveRecord::Base
+  def self.find_by(*args)
```
I think overriding `find_by` can lead to confusing behavior especially since will never return `nil`.

What about making a new method?

```ruby
def self.find_by_or_null(*args)
  find_by(*args) || NullServiceProviderRequest.new
end
```

monfresh replied:
Sure.

zachmargolis commented on spec/controllers/openid_connect/authorization_controller_spec.rb:
```diff
@@ -99,16 +99,21 @@
 
     context 'user is not signed in' do
       it 'redirects to login' do
```
update the comment?

`it 'redirects to login with the request_id'` or something?

monfresh replied:
Yes, will do.

zachmargolis commented on spec/support/features/session_helper.rb:
```diff
@@ -213,5 +213,91 @@ def session_store
+
+      enter_2fa_code
+
+      # expect(page).to have_css('img[src*=sp-logos]')
```
is this commented out on purpose?

monfresh replied:
Yes. I noticed that we currently do not preserve the branded experience on the recovery code page. I opened an issue asking if that was intentional or not but haven't heard back yet. I would assume that we would show the branded nav bar throughout the whole creation process.

monfresh replied:
Yep. In my previous comment I said that I opened an issue already. https://github.com/18F/identity-private/issues/1692

zachmargolis commented on app/controllers/concerns/fully_authenticatable.rb:
```diff
@@ -0,0 +1,14 @@
+
+    return prompt_to_set_up_2fa unless current_user.two_factor_enabled?
+
+    prompt_to_enter_otp
```
this line is red from codeclimate, do we not have test coverage?

monfresh replied:
Probably not for this scenario, which is where a user makes a request from an SP while they are already signed in with email and password only and haven't entered their OTP yet. I'll add a test.

jessieay replied:
link to commit with spec?

monfresh replied:
https://github.com/18F/identity-idp/pull/1265/commits/6e1ff63ab3bd5d21d4aa25b01296392033d1836e

jessieay replied:
darn looks like that was squashed?

monfresh replied:
Here you go: https://github.com/18F/identity-idp/pull/1265/commits/d1ba3b9ee598cd570b06377973e176391674666e
Scroll down to loa1_sso_spec

monfresh commented:
Thanks for the feedback. I addressed it in second commit. PTAL. I'll also fix conflicts with master and push again.

jessieay commented on app/controllers/openid_connect/authorization_controller.rb:
```diff
@@ -84,10 +105,26 @@ def session_params
-                       show_start_page: true }
+      return if sp_session[:request_id]
+
+      session[:sp] = {
```
now that we have the `sp` data in the database, why do we need to store anything other than the UUI in the session?

monfresh replied:
Good question. For several reasons.

- We need to store the request URL to make sure that existing users who sign in are redirected back to the SP. It's easier and faster to query the session rather than the DB, and easier than making sure the request_id is available on all forms and URLs throughout the auth process.
- The branded nav bar depends on the presence of either the issuer in the session or the request_id in the params. In the scenario where a user is signing in, we don't include the request_id params throughout the sign in process because that's only necessary for scenarios where users are switching browsers. Therefore, we need the issuer in the session.
- We have several places that check for `sp_session[:loa3]` in order to do stuff. In order to replace that, we'd need to pass in the `request_id` on every form and URL, and it would also mean a lot more DB lookups.

pkarman replied:
I think @monfresh has taken the right approach. Populating the session once from the db is cleaner, since there is already a single link between the session and the browser (the session cookie).

monfresh commented on spec/views/devise/sessions/new.html.slim_spec.rb:
```diff
@@ -26,7 +26,7 @@
     expect(rendered).
       to have_link(
-        t('links.create_account'), href: sign_up_email_path
+        t('links.create_account'), href: sign_up_email_url(request_id: nil)
```
I believe it needs to be `_url` to include params. Plus, I think we're moving to using `_url` exclusively. I would assume this would fail if it had to be `_path`.

jessieay replied:
We are moving to `_url` for redirects since that is in line with the HTTP spec. AFAIK, you don't need `_url` to include params/

pkarman replied:
my memory is that `_url` came up only because the HTTP spec requires it for Redirect responses. That's all. I agree that `_path` is fine for href values etc, since those are all client-side anyway.

monfresh replied:
I just tried changing to `_path` and it failed. `_url` is required if you want to check for params as well.

monfresh commented on spec/models/user_spec.rb:
```diff
@@ -11,10 +11,10 @@
   end
 
-  it 'should only send one email during creation' do
+  it 'does not send an email when #create is called' do
```
It is very related, and it was intentional. As mentioned in my commit message, Devise automatically calls `send_confirmation_instructions` when the user is created (via a model callback). Unfortunately, Devise does not make it possible to send it additional parameters that can be included in the email. Therefore, I had to create a new `send_custom_confirmation_instructions` method that accepts an argument, and at the same time override Devise's `send_confirmation_instructions` to do nothing. Otherwise, the user would receive 2 emails.

monfresh commented on spec/support/features/session_helper.rb:
```diff
@@ -213,5 +213,91 @@ def session_store
       config.session_store.new({}, config.session_options)
     end
+
+    def sign_up_and_2fa_loa1_user_who_came_from_sp
```
Yep, I knew you would bring this up, but thought I'd get this out for review and then clean it up.

jessieay replied:
you know me soooo well 🥇

monfresh commented on spec/features/saml/loa3_sso_spec.rb:
```diff
@@ -5,20 +5,27 @@
       visit saml_authn_request
-      visit sign_up_email_path
-      user = sign_up_and_set_password
+      click_link t('experiments.demo.get_started')
```
Agreed

monfresh commented on db/migrate/20170321170516_create_service_provider_requests.rb:
```diff
@@ -0,0 +1,14 @@
+
+      t.timestamps null: false
+
+      t.index :uuid
```
Good call

monfresh commented on app/views/sign_up/registrations/show.html.slim:
```diff
@@ -40,8 +41,5 @@
           p.mb3 = link_to \
             t('devise.registrations.start.learn_more'), MarketingSite.help_url, target: '_blank'
-  .center
-    - ab_test(:demo) do |variant, _|
```
This is not removing the A/B testing entirely. It is removing a duplicate div that was causing the LOA3 spec to fail because it found 2 elements matching the "Get started" button. I suppose I could edit the spec to use the first match and then remove this in a separate PR. What do you think?

jessieay replied:
separate might be nice since this PR is quite large as-is, but up to you!

jessieay commented:
Hey @monfresh I just pulled this down locally to test.

CHROME:
- visit http://localhost:3000/test/saml
-  entered my email 

Firefox:
- confirmed email
- set password
- confirmed phone
- confirmed recovery code

after this flow, I was redirected to my profile page rather than the SP, which makes me think that the SP-related info was not being saved properly in Firefox. Am I doing something wrong?

monfresh commented on spec/features/saml/loa1_sso_spec.rb:
```diff
@@ -3,28 +3,35 @@
       click_on t('forms.buttons.continue_to', sp: 'Your friendly Government Agency')
 
       expect(current_url).to eq saml_authn_request
+      expect(ServiceProviderRequest.from_uuid(sp_request_id)).
```
Not that I can think of. I'm open to ideas if you have them. I want to be able to isolate the creation of the ServiceProviderRequest and then make sure it gets deleted. In a controller spec, the entire action will run in one step. AFAIK, I can't use a controller spec to verify that a particular ServiceProviderRequest was created since it will get deleted after calling the action. This controller is special in that it gets called twice, but we only want some of what it does to be called once. I think this is best tested with a feature spec.

jessieay replied:
maybe a request spec? I usually think of feature specs as "end user" focused, so no database-related expectations. Request specs are more backend-y so perhaps this would work there?

monfresh replied:
It might, but I'm not sure it's worth writing a whole new spec that in essence duplicates what the feature spec does, just to test a single expectation.

pkarman replied:
I would expect to see a feature test using `perform_in_browser` to show the full end-to-end multiple browser interplay, from the user's perspective. Am I missing seeing that?

monfresh replied:
As, yes, thanks for reminding me. I meant to add a step that deletes the session. Doing that now.

monfresh replied:
Fixed in https://github.com/18F/identity-idp/pull/1265/commits/c630146a883f21902dde83d0a66c4ce323f98345
PTAL

monfresh commented:
@jessieay You need to come from the sp-rails app. The `test/saml` endpoint doesn't call `SamlIdpController#auth`, which is where all the magic happens.

monfresh commented:
Here are the scenarios I could think of and wrote tests for. Please see if I missed any.

- User comes from SP, clicks "Get started", submits form with invalid email => page displays inline error as before, and URL still contains `request_id` param, and the form still contains the correct request_id in a hidden field.

- User comes from SP, clicks "Get started", submits form with valid email => user is redirected to `/sign_up/verify_email` and receives an email with a link containing the correct request id in a param named `_request_id` (note the underscore prefix) and the order of the params should be `_request_id`, then `confirmation_token`.
  - User clicks "Resend" on the `/sign_up/verify_email` page. => User receives another confirmation email with the same correct params in the link
  - User clicks "try a different email" on the `/sign_up/verify_email` page, then enters a different email => User receives an email with the correct link
  - User clicks the link in the email => User is confirmed and is redirected to `/sign_up/enter_password` with the request_id in the params
  - User sits idle until session times out => SP branding should remain and continuing the account creation should redirect back to SP (I just thought of this new scenario and the second part is failing)

- User comes from SP, clicks "Get started", submits form with valid email, opens email link in a different browser, disables JS in the browser, then enters an invalid password => page displays inline error as before, and URL still contains `request_id` param, and the form still contains the correct request_id in a hidden field.
  - Enter correct password and complete the account creation => you should be redirected to SP

- User comes from SP, clicks "Get started", then stays idle until session times out => branded nav should remain visible, and completing the account creation should redirect back to SP

monfresh commented:
@jessieay I believe I addressed your feedback. We need to get this in to the RC so it can be deployed and tested. If you find anything else, let me know and I'll address it in a separate PR. Thanks!

pkarman commented on spec/support/features/session_helper.rb:
```diff
@@ -257,6 +257,8 @@ def sign_up_and_2fa_loa1_user_who_came_from_sp
       expect(current_url).to eq sign_up_verify_email_url(request_id: sp_request_id, resend: true)
       expect(page).to have_css('img[src*=sp-logos]')
 
+      delete_sp_info_from_session_to_simulate_user_switching_browsers
```
why not actually switch browsers? `perform_in_browser` lets you do that.

monfresh commented:
Used Perform in browser here: https://github.com/18F/identity-idp/pull/1265/commits/4e613570d7793dc45d6351fadbd68b12d03aff8a

monfresh commented on app/views/devise/mailer/confirmation_instructions.html.slim:
```diff
@@ -11,13 +11,16 @@ table.button.expanded.large.radius
                 center
                   = link_to t('mailer.confirmation_instructions.link_text'), \
-                    sign_up_create_email_confirmation_url(confirmation_token: @token), \
+                    sign_up_create_email_confirmation_url(_request_id: \
```
Please read the initial commit message 😄

jessieay replied:
to recap for others:

> Note that the request_id is prefixed with an underscore in the email to
make sure that the URL ends with the confirmation token. By default,
Rails sorts params in link_to helpers alphabetically, but we don't
want request_id to come last because if the user chooses to copy the
link instead of clicking it, and if they don't copy it correctly, the
app won't find the ServiceProviderRequest, whereas an invalid
confirmation token will display an error to the user.

I find this logic sensible, although I am wondering how we can communicate the reasoning for this underscore to future people / our future selves, since it does feel a bit mysterious when browsing the code.

monfresh replied:
Good point. I typically use `git log -S` or I look up the file in question in GitHub and look at the history to find out which commit introduced something and then I look up the PR, assuming the commit message explains the change since we've been pretty good about doing that.

jessieay replied:
Yeah I think that is ideal, but that cannot always be depended on. I will noodle on ways we can make this more obvious, perhaps through some specs

monfresh commented on app/controllers/sign_up/email_confirmations_controller.rb:
```diff
@@ -28,9 +28,30 @@ def process_successful_confirmation
+    end
+
+    def current_sp
+      @_current_sp ||= ServiceProviderRequest.from_uuid(params[:_request_id])
```
Yes, `current_sp_request`. Good catch.

monfresh replied:
Fixed here: https://github.com/18F/identity-idp/pull/1265/commits/cdf1da968ae7405e9216d3c5d21cdfcc65e8242a
PTAL

monfresh commented on app/controllers/saml_idp_controller.rb:
```diff
@@ -38,6 +36,20 @@ def logout
+    needs_idv = identity_needs_verification?
+    analytics.track_event(Analytics::SAML_AUTH, @result.to_h.merge(idv: needs_idv))
+
+    yield needs_idv
```
Because this method is meant to be called with a block so that the `auth` method can `return` if the user `needs_idv`.

monfresh commented on app/controllers/sign_up/registrations_controller.rb:
```diff
@@ -18,6 +18,7 @@ def new
       @register_user_email_form = RegisterUserEmailForm.new
       session[:sign_up_init] = true
       analytics.track_event(Analytics::USER_REGISTRATION_ENTER_EMAIL_VISIT)
+      render :new, locals: { request_id: nil }
```
Where does `nil.to_s` come from? I want to pass in a nil value to the form. When the form renders, the `request_id` hidden input field won't have a value at all.

jessieay replied:
anything in the view is going to inherently have `to_s` called on it. So this will render in the hidden input field as `''`, I believe

monfresh replied:
That doesn't seem to be the case for `nil`:
```html
<input class="block col-12 field hidden" type="hidden" name="user[request_id]" id="user_request_id">
```

It works either way, but I think `nil` conveys the intent better, which is that by default, we don't want any value at all, not even an empty string, because it can imply presence. For example, if we had a conditional in the view that asked `if request_id`, then it would pass with an empty string, but not if it was `nil`.

monfresh commented:
Thanks for the thorough review, @jessieay! Is this good to go now?

jessieay commented on app/models/user.rb:
```diff
@@ -133,4 +133,20 @@ def downcase_keys
+    # no-op
+  end
+
+  # This is basically Devise's send_confirmation_instructions method copied
```
thoughts on my comment about this comment, @monfresh  ?

monfresh replied:
Yes, just remembered: https://github.com/18F/identity-idp/pull/1265/commits/03a4683ea9e053810d6548d8879db923d615c782

monfresh replied:
Thoughts?

jessieay replied:
looks good!

jessieay commented on spec/support/features/session_helper.rb:
```diff
@@ -213,5 +213,143 @@ def session_store
       config.session_store.new({}, config.session_options)
     end
+
+    def sign_up_user_from_sp_without_confirming_email(email)
```
so is the plan to break this up in a future PR? I commented about doing that before and I remember that you replied but the thread is gone now...

jessieay replied:
Oh I see you *did* break it up...WOW still so long.
